### ISSUE TITLE
Provide a way to customize the URL structure

### ISSUE THREAD

**Issue by steve-chavez** [PostgREST/postgrest#1909]
### Problem

We recommend customizing the URL structure [through Nginx](https://postgrest.org/en/v8.0/admin.html#alternate-url-structure), however doing so is complicated and requires additional tools. This can be seen on [ppKrauss/PostgREST-writeAPI](https://github.com/ppKrauss/PostgREST-writeAPI), which uses [mustache templating](https://github.com/ppKrauss/PostgREST-writeAPI/blob/master/nginx-tpl/tpl01-baseBreak.mustache) plus a [nodejs script](https://github.com/ppKrauss/PostgREST-writeAPI/blob/master/writeApi.js) to generate the Nginx rewrite config.

Additionally, changing the URL in the proxy will not be reflected in our OpenAPI spec - which would need rewriting on the proxy as well, which is also more work.

Relying solely on the proxy for this **hurts our extensibility**, ideally PostgREST should be able to adapt to other standards, like more classical REST with nested routes or maybe more complex ones like [Odata](https://www.odata.org/getting-started/basic-tutorial/); and do so outside of PostgREST core.

Note that we do have a way forward to customize the response body(see [this discussion regarding JSON-LD](https://github.com/PostgREST/postgrest/discussions/1790)), but not the URL structure.

### Proposal

Provide a way to customize the URL through the config file or through postgrest-contrib. It could be like so:

```sql
select pgrst.url_rewrites(
  -- more classic REST
  -- /persons/1
  'persons/${value}', 'people?id=eq.${value}'
  -- /persons/1/companies
, 'persons/${value}/companies', '/people?select=*,companies(*)&id=eq.${value}'
  -- bring an RPC to root endpoint
, 'articles/${year}', 'rpc/get_articles?year=${year}'
  -- special route
, 'christmas_2021', 'rpc/get_articles?year=2021&month=december'
);
```

### Related

- https://github.com/PostgREST/postgrest/issues/1086

---

@wolfgangwalther I think you had an idea about a "routing refactor", this may be related.

**Comment by wolfgangwalther** [comment]
> @wolfgangwalther I think you had an idea about a "routing refactor", this may be related.

Absolutely, I had customizing of the URLs for endpoints in mind with that, too.

**Comment by gautam1168** [comment]
@steve-chavez so this would be a separate module that somehow plugs in before the `userApiRequest` function in `ApiRequest.hs` file depending on a configuration? Is there an example of such a module that I can look at to get an idea of how to write such a thing?

**Comment by steve-chavez** [comment]
> Is there an example of such a module that I can look at to get an idea of how to write such a thing?

Hm no, there isn't. This is mostly a high-level idea at this stage, extensibility is pretty new. 

> so this would be a separate module that somehow plugs in before the userApiRequest function in ApiRequest.hs file

Sounds about right, though TBH I haven't looked yet at the implementation details.

**Comment by wolfgangwalther** [comment]
> Sounds about right, though TBH I haven't looked yet at the implementation details.

At a higher level, my idea was roughly along the following lines:
- Currently, the request is parsed in `ApiRequest` by pulling together lots of data from various sources (request itself, config, schema cache) - all this happens for each request, where the implicit routing table is basically built dynamically per request.
- I would like to change that to create a `Router` type that holds all the information for available routes, that can be constructed from the config and schema cache - independent of the actual request.
- When the request comes in, the request can then be matched against the `Router` to find the route it should be handled from.
- This allows to customize the `Router` easily. After creation of the base router from config and schema cache, additional options can be parsed to customize all kinds of things, e.g. the endpoints, accepted media-types, etc... - the suggested `pgrst.url_rewrites` would hook in right there.

Does that make sense?

**Comment by gautam1168** [comment]
Hmm, let me try to echo back what you are saying. There would be a new data type called `Router`. And router and a request can be combined using `applyRouting` function to reform the paths in the incoming `Request` to create a new `Request` with modified path. Maybe something like the following:

```
data Router = Router {
-- some kind of representation of appconfig and dbstructure  
}

makeRouter :: AppConfig -> DbStructure -> Router
applyRouting :: Router -> Request -> Request
userApiRequest :: Router -> Request -> RequestBody -> Either ApiRequestError ApiRequest
```

**Comment by wolfgangwalther** [comment]
Hm. Using the names you suggested, I think `userApiRequest` would be replaced by `makeRouter` and `applyRouting`. Currently `userApiRequest` does both. So something like this maybe:

```haskell
data Router = Router {
-- some kind of representation of appconfig and dbstructure  and routing rules
}

makeRouter :: AppConfig -> DbStructure -> Router
applyRouting :: Router -> Request -> RequestBody -> Either ApiRequestError ApiRequest
```

**Comment by gautam1168** [comment]
@wolfgangwalther yes this makes sense. However, if I try to write a function in haskell whose job is to `run` some rules provided in the configuration file as mentioned in the description of this issue, then I think there are two ways it can be bad:

1. The users specify some rule that is not satisfiable by the function and they do not know that they have made that mistake until runtime. Because there is no `compiler` checking this dsl we are inventing inside the configuration file to specify routing.
2. We end up writing a very extensive module to `check` that the router configuration is sound in every way before we start postgrest (which feels like a big project to me, depending on how complex we make this dsl). And then there could be feature requests on this module making it even more complex and hard to maintain.

So going with postgrest's motto of doing one thing well, I suggest we somehow provide this `configuration` of the router as a program.

 So, something like, if you provide an executable in the configuration file, postgrest will start that executable on a separate process. And we provide a default executable (router) with postgrest which may implement what we have discussed above. For example:

```
# config file
router = "default" # it can be a path to some executable
router-configs = "./routerconfigs"
```

```
# routerconfigs
port = 8080
host = localhost
target-port = 3000 # postgrest is running here
target-host = localhost
```

Now when you start postgrest, and it finds a `router` configuration it will start that program on a separate process. 

This way, we can still allow people to configure routes on postgrest out of the box. And we give them a path to either go with some kind of cloud native architecture like AWS Apigateway or their own kubernetes setup with nginx gradually. We also give them the ability to configure much more than just the routes. Because  this program can just rewrite the entire request, swapping out the request method, body and headers in arbitrary ways that the users can imagine. (This is possible inside the postgrest function we discussed above as well, however, the configuration dsl will become evil at that point).

So basically, Im saying providing a very simple dsl for routing is not going to be enough. And if the goal of postgrest is to run in clouds then we should probably think of how to make people's lives easy when they deploy to a cloud using that cloud's native machinery. So, we can make things that make it dead simple for people to deploy and configure things in AWS, GCP and Azure or at we could invest in making postgrest have a router out of the box when deploying using kubernetes or using a docker compose. And when none of these are available we provide a mock of some equivalent architecture so people can easily spin up and test it in their locals.

**Comment by steve-chavez** [comment]
> Because there is no compiler checking this dsl we are inventing inside the configuration file to specify routing.
> We end up writing a very extensive module to check that the router configuration is sound in every way before we start postgrest (which feels like a big project to me, depending on how complex we make this dsl).

True, it's looking complex. And the DSL I put above only considers the URL part, ideally we'd also be able to extend PostgREST to be compatible with IDL-based standards such as graphql — for this we'd need to parse the request body + map it to PostgREST query params.

That is one of the reasons I mentioned the embedded scripting language idea in https://github.com/PostgREST/postgrest/issues/1698#issuecomment-907922983.

There might be another option. I've just read about [List-based parser combinators](https://wimvanderbauwhede.github.io/articles/list-based-parser-combinators/), which is a way to do parser combinators dinamically. With this we could(in theory) allow to modify PostgREST parser combinators through the configuration file.

We'd need a library for the List-based parser combinators, because there's only an [example for Haskell](https://github.com/wimvanderbauwhede/list-based-combinators-hs). Then we could use that lib for parsing the body + the URL part(maybe).

**Comment by wolfgangwalther** [comment]
>  1. The users specify some rule that is not satisfiable by the function and they do not know that they have made that mistake until runtime. Because there is no `compiler` checking this dsl we are inventing inside the configuration file to specify routing.

Proper strong haskell typing should avoid that.

>  2. We end up writing a very extensive module to `check` that the router configuration is sound in every way before we start postgrest (which feels like a big project to me, depending on how complex we make this dsl). And then there could be feature requests on this module making it even more complex and hard to maintain.

I don't think this is going to be a very complex thing. I'm pretty sure it should be straight-forward to add a parser that reads in a JSON config for the router, maps that to a haskell data type and throws decent parsing errors.

> So going with postgrest's motto of doing one thing well, I suggest we somehow provide this `configuration` of the router as a program.

I think this is just going to complicate things a lot more. For everything that goes beyond the simple set of features that we want to support directly at the PostgREST level, nginx as a reverse-proxy is still the best solution. We don't need to reinvent the wheel here.

> True, it's looking complex. And the DSL I put above only considers the URL part, ideally we'd also be able to extend PostgREST to be compatible with IDL-based standards such as graphql — for this we'd need to parse the request body + map it to PostgREST query params.
> 
> That is one of the reasons I mentioned the embedded scripting language idea in #1698 (comment).

The whole scripting language idea sounds promising - as a second step after the routing refactor. In that second step, we would basically rewrite the request parser in that scripting language, to make it possible to swap it out by a replacement. Does that make sense?

**Comment by gautam1168** [comment]
Hmm yes. The embedded scripting language part gets me more excited than this refactor 😛. And the idea of extending postgrest to be compatible with graphql is very exciting. I would love to see some specs about how to do these things. But for a starter I can refactor the router out as @wolfgangwalther suggested above. 

Can @steve-chavez or @wolfgangwalther throw some more light on how it can be evolved to support graphql or start another issue/rfc for that?

**Comment by steve-chavez** [comment]
> throw some more light on how it can be evolved to support graphql or start another issue/rfc for that?

@gautam1168 I don't have anything specific but really that part should be left for another repo. My main motivation  here is to provide a generic interface for extensibility.

Also, seems that dhall might have issues with graphql because it doesn't have recursiveness?
https://discourse.dhall-lang.org/t/dhall-as-an-alternative-to-openapi-graphql/344/4. Hm, that might be a blocker for us to adopt Dhall.

**Comment by gautam1168** [comment]
hmm ok. Let me open a PR for this refactor first.

**Comment by steve-chavez** [comment]
Regarding embedded scripting languages, there's also [Embedded SQL](https://en.wikipedia.org/wiki/Embedded_SQL) - PostgreSQL has [ECPG](https://www.postgresql.org/docs/13/ecpg-concept.html) for example.

For Haskell, I found [hesql](https://hackage.haskell.org/package/hesql)(does a subset of Postgres' SQL), it doesn't seem maintained but the idea of embedded SQL in PostgREST seems promising.

**Comment by steve-chavez** [comment]
> Also, seems that dhall might have issues with graphql because it doesn't have recursiveness?
https://discourse.dhall-lang.org/t/dhall-as-an-alternative-to-openapi-graphql/344/4. Hm, that might be a blocker for us to adopt Dhall.

I've seen that Dhall supports [toml](https://github.com/dhall-lang/dhall-haskell/tree/master/dhall-toml), that doesn't seem that different from supporting graphql.

So maybe not all things can be done in pure dhall, but dhall can be extended with Haskell. So we'd need to upgrade our Dhall dependency in some cases, but that's not too bad, we'd still support extensibility off the core repo.

> For Haskell, I found hesql(does a subset of Postgres' SQL), it doesn't seem maintained but the idea of embedded SQL in PostgREST seems really cool

That might be shooting for the stars for now, Dhall is looking like the most feasible alternative.

**Comment by steve-chavez** [comment]
There was a request in [gitter](https://gitter.im/begriffs/postgrest?at=621e25358db2b95f0a47984d) about making PostgREST read-only. Of course a read replica or having a read-only role would work but maybe we can allow to restrict our http methods so only GETs are possible and POSTs or other methods don't even touch the db.

Or for a more "db as source of truth" approach, we could reject http methods without hitting the db based on the role's privileges. This requires more smartness from the schema cache though.

**Comment by steve-chavez** [comment]
This discussion derailed a bit of its original purpose. There's a more plausible alternative now on https://github.com/PostgREST/postgrest/issues/1086#issuecomment-1513910590

Basically by defining the custom URL as a view or function name:

```sql
CREATE VIEW "./projects/:id" AS
SELECT * 
FROM projects
WHERE id = current_setting('request.path.id`, false);
```

Let's follow up there.
